src/rpc/blockchain.cpp

```diff
@@ -2774,44 +2790,52 @@ static RPCHelpMan dumptxoutset()
         // would be classified as a block connecting an invalid block.
         disable_network = std::make_unique<NetworkDisable>(connman);
 
-        // Note: Unlocking cs_main before CreateUTXOSnapshot might be racy
-        // if the user interacts with any other *block RPCs.
         invalidate_index = WITH_LOCK(::cs_main, return node.chainman->ActiveChain().Next(target_index));
         InvalidateBlock(*node.chainman, invalidate_index->GetBlockHash());
-        const CBlockIndex* new_tip_index{WITH_LOCK(::cs_main, return node.chainman->ActiveChain().Tip())};
+    }
 
+    Chainstate* chainstate;
+    std::unique_ptr<CCoinsViewCursor> cursor;
+    CCoinsStats stats;
+    UniValue result;
+    UniValue error;
+    {
+        LOCK(node.chainman->GetMutex());
+        chainstate = &node.chainman->ActiveChainstate();
         // In case there is any issue with a block being read from disk we need
         // to stop here, otherwise the dump could still be created for the wrong
         // height.
         // The new tip could also not be the target block if we have a stale
         // sister block of invalidate_index. This block (or a descendant) would
         // be activated as the new tip and we would not get to new_tip_index.
-        if (new_tip_index != target_index) {
-            ReconsiderBlock(*node.chainman, invalidate_index->GetBlockHash());
-            throw JSONRPCError(RPC_MISC_ERROR, "Could not roll back to requested height, reverting to tip.");
+        if (target_index != chainstate->m_chain.Tip()) {
+            error = JSONRPCError(RPC_MISC_ERROR, "Could not roll back to requested height, reverting to tip.");
+        } else {
+            std::tie(cursor, stats, tip) = PrepareUTXOSnapshot(*chainstate, node.rpc_interruption_point);
         }
     }
 
-    UniValue result = CreateUTXOSnapshot(
-        node, node.chainman->ActiveChainstate(), afile, path, temppath);
-    fs::rename(temppath, path);
-
+    if (error.isNull()) {
+        result = WriteUTXOSnapshot(*chainstate, cursor.get(), &stats, tip, afile, path, temppath, node.rpc_interruption_point);
+        fs::rename(temppath, path);
+    }
     if (invalidate_index) {
```

Comment:
In commit "rpc, refactor: Prevent potential race conditions in dumptxoutset" (0c1238917a5edf6bdcae15aab5f91b1bdd2540b9)

Would it make any sense to call `ReconsiderBlock()` and `disable_network.reset()` before calling `WriteUTXOSnapshot` instead of after?

Not sure if that might make the node more usable while the snapshot is being created, or just slow down writing the snapshot and increase resource usage.

Change wouldn't be appropriate in this commit anyway, but could be a possible followup or possible future `dumptxoutset` RPC option if it makes sense.

Reply:
I did some light testing of this and didn't see a negative performance impact. Actually the three times I ran with this change were slightly faster on average than with the current code here. But there was also quite a bit of volatility so that might have been pure luck. It doesn't look like it slows things down though, at least not on my machine.

I have left it out for now but can add it. I am curious what you mean by having it as an RPC option. This seems like something that we should just do if it's safe and we think it's beneficial for users and if not we just leave it out. I'm not sure which users would care enough about such a setting. This is also not a RPC that a lot of people will use when they do it's rather frequently I assume.

Reply:
It seems conceptually simpler to first write the snapshot, reconsider all blocks that were marked invalid and only then restore network activity.

>  make the node more usable while the snapshot is being created

Lower usability is probably feature. The user should not be doing anything important until the chain is back at the original tip, e.g. loading a wallet could still cause confusion.

Reply:
re: https://github.com/bitcoin/bitcoin/pull/29553#discussion_r1725892812

Interesting results and I agree it makes sense to leave this alone for now. I thought it adding an option might be reasonable because doing this might increase resource usage even if it is faster. And maybe it is better if node is unusable for longer as Sjors says. I don't have a clear sense of what is best here.